src/lightning/data/streaming/dataset_optimizer.py

```diff
@@ -790,6 +819,16 @@ def _upload_index(self, cache_dir: str, num_nodes: int, node_rank: Optional[int]
                 node_index_filepath = os.path.join(cache_dir, os.path.basename(remote_filepath))
                 if obj.scheme == "s3":
                     obj = parse.urlparse(remote_filepath)
+                    # Wait for the index to exist on s3.
+                    while True:
+                        try:
+                            s3.head_object(Bucket=obj.netloc, Key=obj.path.lstrip("/"))
+                            break
+                        except Exception as e:
+                            print(e)
```

Comment:
Wouldn't it be better to output the full traceback here? Right now you are only printing the exception, but that won't tell us where it came from, right?

Reply:
The output can be ignored as long it is associated with Not Found. It is expected some ranks can take more time to generate their own index.